File: VaavudCore/LogInViewController.m
//
//  LogInViewController.m
//  Vaavud
//
//  Created by Thomas Stilling Ambus on 12/02/2014.
//  Copyright (c) 2014 Andreas Okholm. All rights reserved.
//

#import "LogInViewController.h"
#import "PasswordUtil.h"
#import "ServerUploadManager.h"
#import "Property+Util.h"
#import "RegisterNavigationController.h"
#import "AccountManager.h"
#import "Mixpanel.h"
#import <FacebookSDK/FacebookSDK.h>

@interface LogInViewController ()<UITextFieldDelegate, UIAlertViewDelegate>

@property (nonatomic, weak) IBOutlet UIView *basicInputView;
@property (nonatomic, weak) IBOutlet UIButton *facebookButton;
@property (nonatomic, weak) IBOutlet UILabel *orLabel;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (nonatomic, weak) IBOutlet UITextField *emailTextField;
@property (nonatomic, weak) IBOutlet UITextField *passwordTextField;
@property (nonatomic) UIAlertView *alertView;

@end

@implementation LogInViewController

BOOL didShowFeedback;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.emailTextField.delegate = self;
    self.passwordTextField.delegate = self;
    
    self.basicInputView.layer.cornerRadius = FORM_CORNER_RADIUS;
    self.basicInputView.layer.masksToBounds = YES;
    
    self.facebookButton.layer.cornerRadius = BUTTON_CORNER_RADIUS;
    self.facebookButton.layer.masksToBounds = YES;

    [self refreshLoginButton];

    if (!self.navigationController || self.navigationController.viewControllers.count <= 1) {
        UIBarButtonItem *item = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"BUTTON_CANCEL", nil) style:UIBarButtonItemStylePlain target:self action:@selector(crossButtonPushed)];
        self.navigationItem.leftBarButtonItem = item;
    }
}

- (void)crossButtonPushed {
    if ([self.navigationController isKindOfClass:[RegisterNavigationController class]]) {
        RegisterNavigationController *registerNavigationController = (RegisterNavigationController *)self.navigationController;
        if (registerNavigationController.registerDelegate) {
            [registerNavigationController.registerDelegate cancelled:self];
        }
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if ([Property isMixpanelEnabled]) {
        [[Mixpanel sharedInstance] track:@"Signup/Login Screen" properties:@{@"Screen": @"Login"}];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    self.alertView.delegate = nil;
    [AccountManager sharedInstance].delegate = nil;
}


- (IBAction)textFieldDidChange:(UITextField *)sender {
    [self refreshLoginButton];
}

- (void)refreshLoginButton {
    NSLog(@"refresh: %d:%d", self.emailTextField.text.length, self.passwordTextField.text.length);

    self.navigationItem.rightBarButtonItem.enabled = self.emailTextField.text.length > 0 && self.passwordTextField.text.length > 0;
}

- (IBAction)doneButtonPushed:(UIBarButtonItem *)sender {
    UIActivityIndicatorView *activityIndicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0, 0, 20, 20)];
    activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleGray;
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:activityIndicator];
    [activityIndicator startAnimating];

    [[AccountManager sharedInstance] registerWithPassword:self.passwordTextField.text email:self.emailTextField.text firstName:nil lastName:nil action:AuthenticationActionLogin success:^(enum AuthenticationResponseType response) {

        [self.passwordTextField resignFirstResponder];
        [self.emailTextField resignFirstResponder];
        self.passwordTextField.delegate = nil;
        self.emailTextField.delegate = nil;
        
        if ([self.navigationController isKindOfClass:[RegisterNavigationController class]]) {
            RegisterNavigationController *registerNavigationController = (RegisterNavigationController *)self.navigationController;
            if (registerNavigationController.registerDelegate) {
                [registerNavigationController.registerDelegate userAuthenticated:(response == AuthenticationResponseCreated) viewController:self];
            }
        }
        
        if (self.completion) {
            self.completion();
        }
    } failure:^(enum AuthenticationResponseType response) {
        if ([Property isMixpanelEnabled]) {
            [[Mixpanel sharedInstance] track:@"Register Error" properties:@{@"Response": [NSNumber numberWithInt:response], @"Screen": @"Login", @"Method": @"Password"}];
        }
        
        [self refreshLoginButton];

        if (response == AuthenticationResponseInvalidCredentials) {
            [self showMessage:NSLocalizedString(@"REGISTER_FEEDBACK_INVALID_CREDENTIALS_MESSAGE", nil) withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_INVALID_CREDENTIALS_TITLE", nil)];
        }
        else if (response == AuthenticationResponseMalformedEmail) {
            [self showMessage:NSLocalizedString(@"REGISTER_FEEDBACK_MALFORMED_EMAIL_MESSAGE", nil) withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_MALFORMED_EMAIL_TITLE", nil)];
        }
        else if (response == AuthenticationResponseLoginWithFacebook) {
            [self showMessage:NSLocalizedString(@"REGISTER_FEEDBACK_ACCOUNT_EXISTS_LOGIN_WITH_FACEBOOK", nil) withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_ACCOUNT_EXISTS_TITLE", nil)];
        }
        else if (response == AuthenticationResponseNoReachability) {
            [self showMessage:NSLocalizedString(@"REGISTER_FEEDBACK_NO_REACHABILITY_MESSAGE", nil) withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_NO_REACHABILITY_TITLE", nil)];
        }
        else {
            [self showMessage:NSLocalizedString(@"REGISTER_FEEDBACK_ERROR_MESSAGE", nil) withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_ERROR_TITLE", nil)];
        }
    }];
}

- (IBAction)facebookButtonPushed:(id)sender {
    [self facebookButtonPushed:sender password:nil];
}

- (void)facebookButtonPushed:(id)sender password:(NSString *)password {
    [self.activityIndicator startAnimating];
    self.facebookButton.titleLabel.hidden = YES;

    didShowFeedback = NO;
    AccountManager *accountManager = [AccountManager sharedInstance];
    accountManager.delegate = self;
    [accountManager registerWithFacebook:password action:AuthenticationActionLogin];
}

- (void)facebookAuthenticationSuccess:(enum AuthenticationResponseType)response {
    [self.activityIndicator stopAnimating];
    self.facebookButton.titleLabel.hidden = NO;

    if ([self.navigationController isKindOfClass:[RegisterNavigationController class]]) {
        RegisterNavigationController *registerNavigationController = (RegisterNavigationController *)self.navigationController;
        if (registerNavigationController.registerDelegate) {
            [registerNavigationController.registerDelegate userAuthenticated:(response == AuthenticationResponseCreated) viewController:self];
        }
    }
    if (self.completion) {
        self.completion();
    }
}

- (void)facebookAuthenticationFailure:(enum AuthenticationResponseType)response
                              message:(NSString *)message
                      displayFeedback:(BOOL)displayFeedback {

    if (LOG_OTHER) NSLog(@"[LogInViewController] error registering user, response=%d, message=%@, displayFeedback=%@", response, message, (displayFeedback ? @"YES" : @"NO"));
    
    [self.activityIndicator stopAnimating];
    self.facebookButton.titleLabel.hidden = NO;

    if ([Property isMixpanelEnabled]) {
        [[Mixpanel sharedInstance] track:@"Register Error" properties:@{@"Response": [NSNumber numberWithInt:response], @"Screen": @"Login", @"Method": @"Facebook"}];
    }
    
    if (displayFeedback && !didShowFeedback) {
        didShowFeedback = YES;
        if (!message || message.length == 0) {
            if (response == AuthenticationResponseEmailUsedProvidePassword) {
                [self promptForPassword];
                return;
            }
            else if (response == AuthenticationResponseNoReachability) {
                [self showMessage:NSLocalizedString(@"REGISTER_FEEDBACK_NO_REACHABILITY_MESSAGE", nil) withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_NO_REACHABILITY_TITLE", nil)];
                return;
            }
            else if (response == AuthenticationResponseFacebookMissingPermission) {
                [self showMessage:NSLocalizedString(@"REGISTER_FEEDBACK_MISSING_FACEBOOK_PERMISSIONS_MESSAGE", nil) withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_MISSING_FACEBOOK_PERMISSIONS_TITLE", nil)];
                return;
            }
            else {
                message = NSLocalizedString(@"REGISTER_FEEDBACK_ERROR_MESSAGE", nil);
            }
        }
        [self showMessage:message withTitle:NSLocalizedString(@"REGISTER_FEEDBACK_ERROR_TITLE", nil)];
    }
}

//- (void)changedEmptiness:(UITextField *)textField isEmpty:(BOOL)isEmpty {
//    UITextField *otherTextField = (textField == self.emailTextField) ? self.passwordTextField : self.emailTextField;
//    if (!otherTextField) {
//        return;
//    }
//    if (!isEmpty && otherTextField.text.length > 0) {
//        self.navigationItem.rightBarButtonItem.enabled = YES;
//    }
//    else {
//        self.navigationItem.rightBarButtonItem.enabled = NO;
//    }
//}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (self.emailTextField.text.length > 0 && self.passwordTextField.text.length > 0) {
        [self refreshLoginButton];
    }
    else if (textField == self.emailTextField) {
        [self.passwordTextField becomeFirstResponder];
    }
    else if (textField == self.passwordTextField) {
        [self.emailTextField becomeFirstResponder];
    }
    return YES;
}

- (void)showMessage:(NSString *)text withTitle:(NSString *)title {
    if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"8.0")) {
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title
                                                                                 message:text
                                                                          preferredStyle:UIAlertControllerStyleAlert];
        
        [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"BUTTON_OK", nil)
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction *action) {}]];
        
        [self presentViewController:alertController animated:YES completion:nil];
    }
    else {
        [[[UIAlertView alloc] initWithTitle:title
                                    message:text
                                   delegate:nil
                          cancelButtonTitle:NSLocalizedString(@"BUTTON_OK", nil)
                          otherButtonTitles:nil] show];
    }
}

- (void)promptForPassword {
    if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"8.0")) {
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"REGISTER_FEEDBACK_ACCOUNT_EXISTS_TITLE", nil)
                                                                                 message:NSLocalizedString(@"REGISTER_FEEDBACK_ACCOUNT_EXISTS_PROVIDE_PASSWORD", nil)
                                                                          preferredStyle:UIAlertControllerStyleAlert];

        [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"BUTTON_CANCEL", nil)
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction *action) {}]];

        [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"BUTTON_OK", nil)
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction *action) {
                                                              UITextField *passwordTextField = alertController.textFields[0];
                                                              if (passwordTextField && passwordTextField.text.length > 0) {
                                                                  [self facebookButtonPushed:nil password:passwordTextField.text];
                                                              }
                                                          }]];
        
        [alertController addTextFieldWithConfigurationHandler:^(UITextField *textField) {
            textField.secureTextEntry = YES;
        }];
        
        [self presentViewController:alertController animated:YES completion:nil];
    }
    else {
        self.alertView = [[UIAlertView alloc] initWithTitle:NSLocalizedString(@"REGISTER_FEEDBACK_ACCOUNT_EXISTS_TITLE", nil)
                                                    message:NSLocalizedString(@"REGISTER_FEEDBACK_ACCOUNT_EXISTS_PROVIDE_PASSWORD", nil)
                                                   delegate:self
                                          cancelButtonTitle:NSLocalizedString(@"BUTTON_CANCEL", nil)
                                          otherButtonTitles:NSLocalizedString(@"BUTTON_OK", nil), nil];
        
        self.alertView.alertViewStyle = UIAlertViewStyleSecureTextInput;
        [self.alertView show];
    }
}

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    if (buttonIndex != alertView.cancelButtonIndex) {
        UITextField *passwordTextField = [alertView textFieldAtIndex:0];
        if (passwordTextField && passwordTextField.text.length > 0) {
            [self facebookButtonPushed:nil password:passwordTextField.text];
        }
    }
}

- (BOOL)alertViewShouldEnableFirstOtherButton:(UIAlertView *)alertView {
    UITextField *passwordTextField = [alertView textFieldAtIndex:0];
    if (passwordTextField && passwordTextField.text.length > 0) {
        return YES;
    }
    return NO;
}

- (void)alertView:(UIAlertView *)alertView didDismissWithButtonIndex:(NSInteger)buttonIndex {
    self.alertView = nil;
}

-(NSUInteger)supportedInterfaceOrientations {
    return UIInterfaceOrientationMaskAll;
}

//- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
//    [self refreshLoginButton];
//    return YES;
//}

@end

File: VaavudCore/LoginViewController.swift
//
//  LoginViewController.swift
//  Vaavud
//
//  Created by Diego R on 11/20/15.
//  Copyright © 2015 Andreas Okholm. All rights reserved.
//

import UIKit
import Firebase

func gotoAppFrom(fromVc: UIViewController, inside parentVc: UIViewController) {
    guard let toVc = UIStoryboard(name: "MainStoryboard", bundle: nil).instantiateInitialViewController() else {
        return
    }
    
    gotoVc(toVc, fromVc: fromVc, parentVc: parentVc)
}

func gotoLoginFrom(fromVc: UIViewController, inside parentVc: UIViewController) {
    let storyboard = UIStoryboard(name: "Login", bundle: nil)
    
    guard let loginNav = storyboard.instantiateInitialViewController() as? UINavigationController else {
        return
    }
    
    
    gotoVc(loginNav, fromVc: fromVc, parentVc: parentVc)
    
}

private func gotoVc(toVc: UIViewController, fromVc: UIViewController, parentVc: UIViewController) {
    parentVc.addChildViewController(toVc)
    fromVc.willMoveToParentViewController(nil)
    
    parentVc.transitionFromViewController(fromVc,
        toViewController: toVc,
        duration: 0.5,
        options: .TransitionFlipFromLeft,
        animations: {}) { _ in
            fromVc.removeFromParentViewController()
            toVc.didMoveToParentViewController(parentVc)
    }
}

class LoginViewController: UIViewController, LoginDelegate {
    @IBOutlet weak var emailField: UITextField!
    @IBOutlet weak var passwordField: UITextField!
    @IBOutlet weak var loginButton: UIBarButtonItem!
    var oldButtonBar: UIBarButtonItem!
    let activityIndicator = UIActivityIndicatorView(frame: CGRectMake(0, 0, 20, 20))
    private let logHelper = LogHelper(.Login)

    
    // MARK: Lifetime
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        logHelper.log("Email")
        
        setupField(emailField)
        setupField(passwordField)
        refreshLoginButton()
    }
    
    
    
    // MARK: User Actions
    
    @IBAction func tappedLogin() {
        if let email = emailField.text, password = passwordField.text {
            activityIndicator.activityIndicatorViewStyle = .White
            oldButtonBar = navigationItem.rightBarButtonItem
            navigationItem.rightBarButtonItem = UIBarButtonItem(customView: activityIndicator)
            activityIndicator.startAnimating()
            navigationController?.view.userInteractionEnabled = false
            view.endEditing(true)
            
            AuthorizationController.shared.login(email, password: password, delegate: self)
        }
        else {
            showError(.Unknown)
        }
    }
    
    // MARK: Segues
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "ForgotPassword", let vc = segue.destinationViewController as? PasswordViewController {
            vc.email = emailField.text
        }
    }
    
    // MARK: Login Delegate
    
    func onSuccess(showActivitySelector: Bool) {
        logHelper.log("Login", properties: ["type":"Success"])

        if showActivitySelector, let vc = self.storyboard?.instantiateViewControllerWithIdentifier("activityVC") {
            navigationController?.interactivePopGestureRecognizer?.enabled = false
            navigationController?.pushViewController(vc, animated: true) {
                self.navigationController?.view.userInteractionEnabled = true
            }
        }
        else {
            gotoAppFrom(navigationController!, inside: view.window!.rootViewController!)
        }
    }
    
    func onError(error: LoginError) {
        logHelper.log("Login", properties: ["type":"Error"])

        dispatch_async(dispatch_get_main_queue(), {
            self.activityIndicator.stopAnimating()
            self.navigationController?.view.userInteractionEnabled = true
            self.navigationItem.rightBarButtonItem = self.oldButtonBar
        })
        
        showError(error)
    }
    
    // MARK: Textfield Delegate
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        if textField == emailField {
            passwordField.becomeFirstResponder()
        }
        else {
            guard validInput() else {
                return false
            }
            
            tappedLogin()
        }
        
        return true
    }
    
    func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
        refreshLoginButton()
        return true
    }
    
    // MARK: Convenience
    
    private func showError(error: LoginError) {
        dispatch_async(dispatch_get_main_queue()) {
            VaavudInteractions().showLocalAlert("LOGIN_ERROR_TITLE", messageKey: error.key, otherKey: "BUTTON_OK", action: {}, on: self)
        }
    }
    
    private func refreshLoginButton() {
        loginButton.enabled = validInput()
    }
    
    private func validInput() -> Bool {
        return (emailField.text?.containsString("@") ?? false) && !passwordField.text!.isEmpty
    }
    
    private func setupField(field: UITextField) {
        let attributes = [NSForegroundColorAttributeName : UIColor.vaavudRedColor().colorWithAlphaComponent(0.3)]
        field.attributedPlaceholder = NSAttributedString(string: field.placeholder ?? "", attributes: attributes)
    }
}

class PasswordViewController: UIViewController, UITextFieldDelegate, LoginDelegate {
    
    var email: String?
    @IBOutlet weak var emailField: UITextField!
    @IBOutlet weak var sendButton: UIButton!
    private var oldButtonBar: UIBarButtonItem!
    private let logHelper = LogHelper(.Login)
    
    // MARK: Lifetime
    
    override func viewDidLoad() {
        logHelper.log("ForgotPassword")

        emailField.text = email
        refreshSendButton()
    }
    
    // MARK: User Actions
    
    @IBAction func tappedSend() {
        let activityIndicator = UIActivityIndicatorView(frame: CGRectMake(0, 0, 20, 20))
        activityIndicator.activityIndicatorViewStyle = .White
        oldButtonBar = navigationItem.rightBarButtonItem
        navigationItem.rightBarButtonItem = UIBarButtonItem(customView: activityIndicator)
        activityIndicator.startAnimating()

        AuthorizationController.shared.resetPassword(emailField.text!, delegate: self)
        navigationController?.view.userInteractionEnabled = false
    }
    
    // MARK: Textfield Delegate
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        guard validInput() else {
            return false
        }
        
        tappedSend()
        
        return true
    }
    
    func onSuccess(showActivitySelector: Bool) {
        logHelper.log("ForgotPassword",properties: ["type":"Success"])

        navigationItem.rightBarButtonItem = oldButtonBar

        VaavudInteractions().showLocalAlert("Thank you", messageKey: "We have sent an email to you with instructions.", otherKey: "BUTTON_OK", action: { [unowned self] in self.goBack() }, on: self)
    }
    
    func goBack() {
        if let nav = navigationController {
            navigationController?.popViewControllerAnimated(true) {
                nav.view.userInteractionEnabled = true
            }
        }
    }
    
    func onError(error: LoginError) {
        logHelper.log("ForgotPassword",properties: ["type":"Error"])
        navigationItem.rightBarButtonItem = oldButtonBar
        navigationController?.view.userInteractionEnabled = true
        VaavudInteractions().showLocalAlert("LOGIN_ERROR_TITLE", messageKey: error.key, otherKey: "BUTTON_OK", action: {
            }, on: self)
    }
    
    func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
        refreshSendButton()
        return true
    }
    
    // MARK: Convenience
    
    private func refreshSendButton() {
        sendButton.enabled = validInput()
    }
    
    private func validInput() -> Bool {
        return emailField.text?.containsString("@") ?? false
    }
}
